proposed/non-primitive-constant-data.md

```diff
@@ -0,0 +1,471 @@
+IL_0008: ldc.i4.3
+IL_0009: call instance void valuetype [System.Memory]System.ReadOnlySpan`1<uint8>::.ctor(void*, int32)
+```
+
+
+- Primitive data for a dynamically constructed array
+```csharp
+void Constants()
+{
+    int[] data = new byte[]{0,1,2,4};
+}
+```
+
+Which is represented in IL using an RVA static field on an anonymous type which is used in conjunction with the System.Runtime.CompilerServices.RuntimeHelpers.InitializeArray function to dynamically create an array based on a set of data which was statically placed into the assembly.
+```
+IL_0001: ldc.i4.4
+IL_0002: newarr [mscorlib]System.Byte
+IL_0007: dup
+IL_0008: ldtoken field int32 '<PrivateImplementationDetails>'::'12DADA1FFF4D4787ADE3333147202C3B443E376F'
+IL_000d: call void [mscorlib]System.Runtime.CompilerServices.RuntimeHelpers::InitializeArray(class [mscorlib]System.Array, valuetype [mscorlib]System.RuntimeFieldHandle)
+IL_0012: stloc.0
+```
+## New valid constant forms
+
+The proposal is to provide a set of CompilerServices apis implemented as intrinsics (in most cases) which will allow the use of more complex constants. The intention is to permit more complex constant forms for ReadOnlySpan<T> access, constant access, and through IL compiler magic, add new forms of array initialization that are not currently supported through the InitializeArray pattern.
+
+### Definition of types which can be represented as constants
+- Must be a struct
+- Generic structures are permitted, and only those generic type parameters which contribute to field layout shall have impact on whether the type can be represented as a constant.
+- The struct may have any form of layout: sequential layout, auto layout, or explicit layout.
+- Without pointers of any form (No object reference, IntPtr, UIntPtr, pointer, function pointer, or TypedReference)
+- The type must be attributed with either WellKnownLayoutTypeAttribute or WellKnownLayoutVectorTypeAttribute
+- All fields must also be of a type which can be represented as a constant
+
+### Well known type layout
+Given that IL binaries may be loaded on many different platforms with different type layout rules, constant data must be represented in a consistent manner across all of them. For consistent type layout, types shall be laid out in precisely a sequential manner as described in metadata, with *no* packing between any fields. For instance, this struct will utilize 9 bytes when stored. In addition, if the type uses explicit layout, even if there are overlapping fields, those fields are not to be overlapped in the WellKnownLayout, the layout of the final form will be based on copying out of the data stored in the PE file in metadata order into the runtime layout. An empty structure will consume 0 bytes of storage, and thus while it may be used as a field within a constant, is not itself permitted to be used to instantiate CreateSpan<T> or LoadConstant<T>
```

Comment:
Mono supports platforms that doesn't support unaligned loads means all `WellKnownLayout` types and its fields must be accessed with `unaligned.` prefix.